libraries/AC_SafeRTL/AC_SafeRTL.cpp

```diff
@@ -0,0 +1,573 @@
+    if (_last_index < _current_path_len - 10) {
+        return true;
+    }
+
+    int16_t potential_amount_to_simplify = _simplification_bitmask.size() - _simplification_bitmask.count();
+
+    // if simplifying will remove more than 10 points, just do it
+    if (potential_amount_to_simplify >= 10) {
+        zero_points_by_simplification_bitmask();
+        remove_empty_points();
+        // end by resetting the state of the cleanup methods.
+        reset_simplification(true);
+        reset_pruning(true);
+        return true;
+    }
+
+    int16_t potential_amount_to_prune = 0;
+    for (int16_t i = 0; i <= _prunable_loops_last_index; i++) {
+        // add 1 at the end, because a pruned loop is always replaced by one new point.
+        potential_amount_to_prune += _prunable_loops[i].end_index - _prunable_loops[i].start_index + 1;
+    }
+
+    // if pruning could remove 10+ points, prune loops until 10 or more points have been removed (doesn't necessarily prune all loops)
+    if (potential_amount_to_prune >= 10) {
+        zero_points_by_loops(10);
+        remove_empty_points();
+        // end by resetting the state of the cleanup methods.
+        reset_simplification(true);
+        reset_pruning(true);
+        return true;
+    }
+
+    // as a last resort, see if pruning and simplifying together would remove 10+ points.
+    if (potential_amount_to_prune + potential_amount_to_simplify >= 10) {
+        zero_points_by_simplification_bitmask();
+        zero_points_by_loops(10);
+        remove_empty_points();
+        // end by resetting the state of the cleanup methods.
+        reset_simplification(true);
+        reset_pruning(true);
+        return true;
+    }
+    return false;
+}
+
+/**
+*   A hard reset will "forget" the optimizations that can be made. A soft reset
+*   should be used when a point is appended, a hard reset should be used when the existing path is altered.
+*/
+void AC_SafeRTL::reset_simplification(bool hard)
+{
+    if (hard) {
+        _simplification_clean_until = 0;
+    }
+    _simplification_complete = false;
+    _simplification_stack_last_index = -1;
+    _simplification_bitmask.set();
+}
+
+/**
+*   A hard reset will "forget" the optimizations that can be made. A soft reset
+*   should be used when a point is appended, a hard reset should be used when the existing path is altered.
+*/
+void AC_SafeRTL::reset_pruning(bool hard)
+{
+    if (hard) {
+        _pruning_clean_until = 0;
+    }
+    _pruning_complete = false;
+    _pruning_current_i = _pruning_clean_until;
+    _pruning_min_j = _pruning_clean_until+2;
+    _prunable_loops_last_index = -1; // clear the loops that we've recorded
+}
+
+void AC_SafeRTL::zero_points_by_simplification_bitmask()
+{
+    for (int16_t i = 0; i <= _last_index; i++) {
+        if (!_simplification_bitmask[i]) {
+            _simplification_clean_until = MIN(_simplification_clean_until, i-1);
+            if (_path[i] != Vector3f(0.0f, 0.0f, 0.0f)) {
```

Comment:
Vector3f::is_zero exists apparently so follow up commit uses it instead of this existing test.